File: lab spring25/assignment3/taskC.cpp
#include <stdio.h>
#define ll long long

ll fastMod(ll a, ll b){
    if(b<=1) return a;

    ll x = fastMod(a,b>>1);
    return (((x*x)%107)*(b&1?a:1))%107;
}

int main(){
    freopen("input.txt","r",stdin);
    ll a,b;
    scanf("%lld %lld",&a,&b);
    printf("%lld\n",fastMod(a%107,b));
}

File: lab spring25/python a3/taskC.py
import sys
sys.stdin = open("input.txt","r")

def fastMod(a,b):
    if b<=1: return a

    x = fastMod(a,b>>1)
    if b&1: return ((x*x)%107*a)%107
    else: return (x*x)%107

a,b = list(map(int,input().split()))
print(fastMod(a%107,b))